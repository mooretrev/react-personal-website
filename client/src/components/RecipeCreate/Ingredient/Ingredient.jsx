import React from 'react';
import PropTypes from 'prop-types';
import TextField from '@material-ui/core/TextField';
import MenuItem from '@material-ui/core/MenuItem';
import Select from '@material-ui/core/Select';
import { Grid } from '@material-ui/core';

function Ingredient(props) {
  // TODO remove these and add input parsing
  const handleIngredientChange = (event) => {
    props.onIngredientChange(event, props.i, props.j);
  };

  const handleSizeChange = (event) => {
    props.onSizeChange(event, props.i, props.j);
  };

  const handleUnitChange = (event) => {
    props.onUnitChange(event, props.i, props.j);
  };

  const {
    i,
    j,
    sizeValue,
    ingredientValue,
    unitValue,
  } = props;

  return (
    <div>
      <Grid container>
        <Grid item xs={4}>
          <TextField value={ingredientValue} id={`ingredient_${i}_${j}`} label="Ingredient" variant="outlined" onChange={handleIngredientChange} />
        </Grid>
        <Grid item xs={4}>
          <TextField value={sizeValue} id={`size_${i}_${j}`} label="Size" variant="outlined" onChange={handleSizeChange} />
        </Grid>
        <Grid item xs={4}>
          <Select
            id={`unit_${i}_${j}`}
            onChange={handleUnitChange}
            value={unitValue}
          >
            <MenuItem value="Cup">Cup</MenuItem>
            <MenuItem value="Tbsp">Tbsp</MenuItem>
            <MenuItem value="Tsp">Tsp</MenuItem>
          </Select>
        </Grid>
      </Grid>
    </div>
  );
}

Ingredient.propTypes = {
  onIngredientChange: PropTypes.func.isRequired,
  onSizeChange: PropTypes.func.isRequired,
  onUnitChange: PropTypes.func.isRequired,
  ingredientValue: PropTypes.string.isRequired,
  sizeValue: PropTypes.string.isRequired,
  unitValue: PropTypes.string.isRequired,
  i: PropTypes.number.isRequired,
  j: PropTypes.number.isRequired,

};

export default Ingredient;
