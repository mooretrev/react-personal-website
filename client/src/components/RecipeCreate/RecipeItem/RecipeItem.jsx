import React from 'react';
import { TextField, Typography } from '@material-ui/core';
import PropTypes from 'prop-types';

function RecipeCreateItem(props) {
  const { value } = props;
  const handleRecipeItemChange = (event) => {
    props.onRecipeItemChange(event, props.index);
  };

  return (
    <div>
      <Typography variant="h3">Create Item</Typography>
      <TextField value={value} onChange={handleRecipeItemChange} id="outlined-basic" label="Recipe Item" variant="outlined" />
    </div>
  );
}

RecipeCreateItem.propTypes = {
  value: PropTypes.string.isRequired,
  onRecipeItemChange: PropTypes.func.isRequired,
  index: PropTypes.number.isRequired,
};

export default RecipeCreateItem;
