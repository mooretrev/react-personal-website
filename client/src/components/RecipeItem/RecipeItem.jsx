import React from 'react';
import Typography from '@material-ui/core/Typography';
import Grid from '@material-ui/core/Grid';
import makeStyles from '@material-ui/core/styles/makeStyles';
import PropTypes from 'prop-types';

const useStyles = makeStyles(() => ({
  text: {
    padding: '4px',
    // margin: '4px'
  },
  unit: {
    textAlign: 'right',
    paddingRight: '7%',
  },
}));

function RecipeItem(props) {
  const classes = useStyles();

  const createIngredients = () => {
    const { ingredients } = props.recipeItem;
    const items = [];
    for (let i = 0; i < ingredients.length; i += 1) {
      items.push(
        <Grid container key={i}>
          <Grid item xs={8}>
            <Typography className={classes.text} variant="h6">
              {ingredients[i].ingredient}
              {' '}
            </Typography>
          </Grid>
          <Grid xs={4}>
            <Typography className={classes.unit} variant="h6">
              {ingredients[i].size}
              {' '}
              {ingredients[i].unit}
              {' '}
            </Typography>
          </Grid>
        </Grid>,

      );
    }
    return items;
  };

  const { recipeItem } = props;

  return (
    <div>
      <Typography className={classes.text} variant="h4">
        {' '}
        {recipeItem.recipe_item}
      </Typography>
      {createIngredients()}
    </div>
  );
}

RecipeItem.propTypes = {
  recipeItem: PropTypes.element.isRequired,
};

export default RecipeItem;
