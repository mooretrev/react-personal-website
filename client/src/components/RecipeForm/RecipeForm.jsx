import React, { useState } from 'react';
import { useAuth0 } from '@auth0/auth0-react';
import Typography from '@material-ui/core/Typography';
import Card from '@material-ui/core/Card';
import TextField from '@material-ui/core/TextField';
import Button from '@material-ui/core/Button';
import Grid from '@material-ui/core/Grid';
import PropTypes from 'prop-types';
import RecipeItem from '../RecipeCreate/RecipeItem/RecipeItem.jsx';
import Ingredient from '../RecipeCreate/Ingredient/Ingredient.jsx';
import CreateRecipe from '../../api/Recipes/CreateRecipe.js';
import EditRecipe from '../../api/Recipes/EditRecipe.js';

function RecipeForm(props) {
  /* STATES */
  const {
    recipeItems,
    ingredients,
    sizes,
    units,
    recipeName,
  } = props;
  const { getAccessTokenSilently } = useAuth0();

  const [recipeItemsState, setRecipeItems] = useState(recipeItems);
  const [ingredientsState, setIngredients] = useState(ingredients);
  const [sizesState, setSizes] = useState(sizes);
  const [unitsState, setUnits] = useState(units);
  const [recipeNameState, setRecipeName] = useState(recipeName);
  const [requestSuccessState, setRequestSuccess] = useState('');

  const resetToIntialState = () => {
    setRecipeName('');
    setRecipeItems(['']);
    setIngredients([['']]);
    setSizes([['']]);
    setUnits([['']]);
  };

  const handleRecipeNameChange = (event) => {
    setRecipeName(event.target.value);
  };

  const handleRecipeItemChange = async (event, i) => {
    const iNum = parseInt(i, 10);
    const newRecipeItems = recipeItemsState.slice();
    newRecipeItems[iNum] = event.target.value;
    setRecipeItems(newRecipeItems);

    if ((iNum + 1) === recipeItemsState.length) {
      newRecipeItems.push('');
      setRecipeItems(newRecipeItems);

      const newIngredients = ingredientsState.slice();
      newIngredients.push(['']);
      setIngredients(newIngredients);

      const newSizes = sizesState.slice();
      newSizes.push(['']);
      setSizes(newSizes);

      const newUnits = unitsState.slice();
      newUnits.push(['']);
      setUnits(newUnits);
    }
  };

  const handleIngredientChange = (event, i, j) => {
    const iNum = parseInt(i, 10);
    const jNum = parseInt(j, 10);

    const newIngredientsT = ingredientsState.slice();
    newIngredientsT[iNum][jNum] = event.target.value;
    setIngredients(newIngredientsT);

    if (jNum + 1 === ingredientsState[iNum].length) {
      const newIngredients = ingredientsState.slice();
      newIngredients[iNum].push('');
      setIngredients(newIngredients);

      const newSizes = sizesState.slice();
      newSizes[iNum].push('');
      setSizes(newSizes);

      const newUnits = unitsState.slice();
      newUnits[iNum].push('');
      setUnits(newUnits);
    }
  };

  const handleSizeChange = (event, i, j) => {
    const iNum = parseInt(i, 10);
    const jNum = parseInt(j, 10);
    const newSizes = sizesState.slice();
    newSizes[iNum][jNum] = event.target.value;
    setSizes(newSizes);
  };

  const handleUnitChange = (event, i, j) => {
    const iNum = parseInt(i, 10);
    const jNum = parseInt(j, 10);
    const newUnits = unitsState.slice();
    newUnits[iNum][jNum] = event.target.value;
    setUnits(newUnits);
  };

  const handleSubmit = async () => {
    const token = await getAccessTokenSilently();
    if (props.edit) {
      EditRecipe(
        token,
        props.id,
        recipeNameState,
        recipeItemsState,
        ingredientsState,
        sizesState,
        unitsState,
      );
      props.history.push(`/recipes/${props.id}`);
    } else {
      CreateRecipe(token, recipeNameState, recipeItemsState, ingredientsState, sizesState, unitsState);
      setRequestSuccess(true);
      resetToIntialState();
    }
  };

  const createCreationMessage = () => {
    if (requestSuccessState === true) {
      return (
        <Grid item xs={12}>
          <Typography variant="subtitle1">
            Recipe created successfully.
          </Typography>
        </Grid>
      );
    }
    return <div />;
  };

  const createIngredient = (i, j) => (
    <Ingredient
      key={`ingredient${j}`}
      onIngredientChange={handleIngredientChange}
      onSizeChange={handleSizeChange}
      onUnitChange={handleUnitChange}
      i={i}
      j={j}
      unitValue={unitsState[i][j]}
      sizeValue={sizesState[i][j]}
      ingredientValue={ingredientsState[i][j]}
    />
  );

  const createIngredients = (i) => {
    const items = [];
    for (let j = 0; ingredientsState[i] !== undefined && j < ingredientsState[i].length; j += 1) {
      items.push(createIngredient(i, j));
    }
    return items;
  };

  const createRecipeItem = (i) => (
    <div>
      <Card>
        <RecipeItem
          value={recipeItemsState[i]}
          index={i}
          onRecipeItemChange={handleRecipeItemChange}
        />
        {createIngredients(i)}
      </Card>
    </div>
  );

  const createElements = () => {
    const items = [];
    for (let i = 0; i < recipeItemsState.length; i += 1) {
      const item = createRecipeItem(i);
      const gridItem = (
        <Grid item xs={12}>
          {item}
        </Grid>
      );
      items.push(gridItem);
    }
    return items;
  };

  return (
    <div>
      <Grid key="form" container spacing={2}>
        <Grid item xs={12}>
          <Card>
            <Typography variant="h2">Create Recipe</Typography>
            <TextField value={recipeNameState} onChange={handleRecipeNameChange} id="outlined-basic" label="Recipe Name" variant="outlined" />
          </Card>
        </Grid>
        {createElements()}
        {createCreationMessage()}
        <Grid item xs={12}>
          <Button onClick={handleSubmit} variant="contained" color="primary">Submit</Button>
        </Grid>
      </Grid>
    </div>
  );
}

RecipeForm.propTypes = {
  recipeItems: PropTypes.element.isRequired,
  ingredients: PropTypes.element.isRequired,
  sizes: PropTypes.element.isRequired,
  units: PropTypes.element.isRequired,
  recipeName: PropTypes.element.isRequired,
  edit: PropTypes.element.isRequired,
  id: PropTypes.element.isRequired,
  history: PropTypes.element.isRequired,

};

export default RecipeForm;
