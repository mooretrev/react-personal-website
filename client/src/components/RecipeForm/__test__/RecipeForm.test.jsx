import React from 'react';
import { shallow, mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import RecipeForm from '../RecipeForm.jsx';
import RecipeItem from '../../RecipeCreate/RecipeItem/RecipeItem.jsx';
import Ingredient from '../../RecipeCreate/Ingredient/Ingredient.jsx';

describe('<RecipeForm />', () => {
  it('should render correctly', () => {
    const recipeForm = (
      <RecipeForm
        id="-1"
        history={0}
        edit={false}
        recipeItems={['']}
        ingredients={[['']]}
        sizes={[['']]}
        units={[['']]}
        recipeName=""
      />
    );
    const wrapper = shallow(recipeForm);
    expect(toJson(wrapper)).toMatchSnapshot();
  });
  it('should render correctly', () => {
    const recipeForm = (
      <RecipeForm
        id="-1"
        history={0}
        edit={false}
        recipeItems={['Dough', 'Bread']}
        ingredients={[['Flour', 'Yeast'], ['Bread Flour', 'Water', 'Sugar']]}
        sizes={[['1', '2'], ['1', '1', '1']]}
        units={[['Cup', 'Cup'], ['Cup', 'Cup', 'Cup']]}
        recipeName=""
      />
    );
    const wrapper = shallow(recipeForm);
    expect(toJson(wrapper)).toMatchSnapshot();
  });
  it('recipe items should increment on recipe item change', () => {
    const recipeForm = (
      <RecipeForm
        id="-1"
        edit={false}
        recipeItems={['']}
        ingredients={[['']]}
        sizes={[['']]}
        units={[['']]}
        recipeName=""
      />
    );
    const wrapper = mount(recipeForm);
    expect(wrapper.find(RecipeItem)).toHaveLength(1);
    const recipeItemInput = wrapper.find('input').at(1);
    recipeItemInput.simulate('change', { target: { value: 'Hello' } });
    expect(wrapper.find(RecipeItem)).toHaveLength(2);
  });
  it('should add another ingredient when needed', () => {
    const recipeForm = (
      <RecipeForm
        id="-1"
        edit={false}
        recipeItems={['']}
        ingredients={[['']]}
        sizes={[['']]}
        units={[['']]}
        recipeName=""
      />
    );
    const wrapper = mount(recipeForm);
    expect(wrapper.find(Ingredient)).toHaveLength(1);
    const recipeItemInput = wrapper.find('input').at(2);
    recipeItemInput.simulate('change', { target: { value: 'Hello' } });
    expect(wrapper.find(Ingredient)).toHaveLength(2);
  });
});
