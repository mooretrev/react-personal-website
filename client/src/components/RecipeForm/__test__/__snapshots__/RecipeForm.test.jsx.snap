// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RecipeForm /> should render correctly 1`] = `
<div>
  <WithStyles(ForwardRef(Grid))
    container={true}
    spacing={2}
  >
    <WithStyles(ForwardRef(Grid))
      item={true}
      xs={12}
    >
      <WithStyles(ForwardRef(Card))>
        <WithStyles(ForwardRef(Typography))
          variant="h2"
        >
          Create Recipe
        </WithStyles(ForwardRef(Typography))>
        <WithStyles(ForwardRef(TextField))
          id="outlined-basic"
          label="Recipe Name"
          onChange={[Function]}
          value=""
          variant="outlined"
        />
      </WithStyles(ForwardRef(Card))>
    </WithStyles(ForwardRef(Grid))>
    <WithStyles(ForwardRef(Grid))
      item={true}
      key="recipeItem0"
      xs={12}
    >
      <div>
        <WithStyles(ForwardRef(Card))>
          <RecipeCreateItem
            index={0}
            onRecipeItemChange={[Function]}
            value=""
          />
          <Ingredient
            i={0}
            ingredientValue=""
            j={0}
            key="ingredient0"
            onIngredientChange={[Function]}
            onSizeChange={[Function]}
            onUnitChange={[Function]}
            sizeValue=""
            unitValue=""
          />
        </WithStyles(ForwardRef(Card))>
      </div>
    </WithStyles(ForwardRef(Grid))>
    <div />
    <WithStyles(ForwardRef(Grid))
      item={true}
      xs={12}
    >
      <WithStyles(ForwardRef(Button))
        color="primary"
        onClick={[Function]}
        variant="contained"
      >
        Submit
      </WithStyles(ForwardRef(Button))>
    </WithStyles(ForwardRef(Grid))>
  </WithStyles(ForwardRef(Grid))>
</div>
`;

exports[`<RecipeForm /> should render correctly 2`] = `
<div>
  <WithStyles(ForwardRef(Grid))
    container={true}
    spacing={2}
  >
    <WithStyles(ForwardRef(Grid))
      item={true}
      xs={12}
    >
      <WithStyles(ForwardRef(Card))>
        <WithStyles(ForwardRef(Typography))
          variant="h2"
        >
          Create Recipe
        </WithStyles(ForwardRef(Typography))>
        <WithStyles(ForwardRef(TextField))
          id="outlined-basic"
          label="Recipe Name"
          onChange={[Function]}
          value=""
          variant="outlined"
        />
      </WithStyles(ForwardRef(Card))>
    </WithStyles(ForwardRef(Grid))>
    <WithStyles(ForwardRef(Grid))
      item={true}
      key="recipeItem0"
      xs={12}
    >
      <div>
        <WithStyles(ForwardRef(Card))>
          <RecipeCreateItem
            index={0}
            onRecipeItemChange={[Function]}
            value="Dough"
          />
          <Ingredient
            i={0}
            ingredientValue="Flour"
            j={0}
            key="ingredient0"
            onIngredientChange={[Function]}
            onSizeChange={[Function]}
            onUnitChange={[Function]}
            sizeValue="1"
            unitValue="Cup"
          />
          <Ingredient
            i={0}
            ingredientValue="Yeast"
            j={1}
            key="ingredient1"
            onIngredientChange={[Function]}
            onSizeChange={[Function]}
            onUnitChange={[Function]}
            sizeValue="2"
            unitValue="Cup"
          />
        </WithStyles(ForwardRef(Card))>
      </div>
    </WithStyles(ForwardRef(Grid))>
    <WithStyles(ForwardRef(Grid))
      item={true}
      key="recipeItem1"
      xs={12}
    >
      <div>
        <WithStyles(ForwardRef(Card))>
          <RecipeCreateItem
            index={1}
            onRecipeItemChange={[Function]}
            value="Bread"
          />
          <Ingredient
            i={1}
            ingredientValue="Bread Flour"
            j={0}
            key="ingredient0"
            onIngredientChange={[Function]}
            onSizeChange={[Function]}
            onUnitChange={[Function]}
            sizeValue="1"
            unitValue="Cup"
          />
          <Ingredient
            i={1}
            ingredientValue="Water"
            j={1}
            key="ingredient1"
            onIngredientChange={[Function]}
            onSizeChange={[Function]}
            onUnitChange={[Function]}
            sizeValue="1"
            unitValue="Cup"
          />
          <Ingredient
            i={1}
            ingredientValue="Sugar"
            j={2}
            key="ingredient2"
            onIngredientChange={[Function]}
            onSizeChange={[Function]}
            onUnitChange={[Function]}
            sizeValue="1"
            unitValue="Cup"
          />
        </WithStyles(ForwardRef(Card))>
      </div>
    </WithStyles(ForwardRef(Grid))>
    <div />
    <WithStyles(ForwardRef(Grid))
      item={true}
      xs={12}
    >
      <WithStyles(ForwardRef(Button))
        color="primary"
        onClick={[Function]}
        variant="contained"
      >
        Submit
      </WithStyles(ForwardRef(Button))>
    </WithStyles(ForwardRef(Grid))>
  </WithStyles(ForwardRef(Grid))>
</div>
`;
